### PR TITLE
Hod clone

### LINKED ISSUE #97: hod clone  - clones a dist in a target directory
We would like to use zipped eggs for deployment since it means shared file systems are abused less on startup (meaning we will have better interactivity). However, if the `etc` dir is zipped it makes it harder for people to make new hodconf distributions based on an existing dist. Therefore, we want a `hod clone` command as follows:

`hod clone --dist=<some-dist> --output-dir=<some-output-dir>`

`--dist` and `--output-dir` are non positional required arguments. This will copy the contents of the `<some-dist>` to `<some-output-dir>`.

### PR DESCRIPTION
**Pull request by ehiggs**
This PR should fix #97.

### CODE REVIEW

**Comment by boegel** [comment]
#100 \o/

**Comment by ehiggs** [comment]
:100:!

File: docs/Command_line_interface.rst [hunk 1]
```diff
@@ -35,6 +35,7 @@ Running ``hod`` without arguments is equivalent to ``hod --help``, and results i
         genconfig       Write hod configs to a directory for diagnostic purposes
         connect         Connect to a hod cluster.
         clean           Remove stale cluster info.
+        clone           Write hod configs to a directory for edtting purposes.
```
Comment (boegel) [inline docs/Command_line_interface.rst]: shouldn't we list these alphabetically?

any other order is arbitrary/confusing?

File: hod/main.py [hunk 2]
```diff
@@ -45,6 +46,7 @@
     connect.ConnectSubCommand,
     clean.CleanSubCommand,
     relabel.RelabelSubCommand,
+    clone.CloneSubCommand,
```
Comment (boegel) [inline hod/main.py]: let's sort these alphabetically...

File: hod/main.py [hunk 3]
```diff
@@ -32,7 +32,8 @@
 import sys
 
 import hod
-from hod.subcommands import batch, connect, clean, create, dists, genconfig, helptemplate, relabel, listcmd
+from hod.subcommands import (batch, connect, clean, create, dists, genconfig,
+        helptemplate, relabel, listcmd, clone)
```
Comment (boegel) [inline hod/main.py]: your alphabet is broken here with the last two (and `connect`) ;)

File: hod/subcommands/clone.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    HELP = "Write hod configs to a directory for editing purposes."
+
+    def run(self, args):
+        """Run 'genconfig' subcommand."""
+        optparser = CloneOptions(go_args=args, usage=self.usage_txt)
+        if len(optparser.args) < 3:
+            sys.stderr.write(self.usage())
+            sys.exit(1)
+
+        dist = optparser.args[1]
```
Comment (boegel) [inline hod/subcommands/clone.py]: hmm, shouldn't we let people use `--dist` here too, like they do with `batch` and `create`?

File: hod/subcommands/clone.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+class CloneSubCommand(SubCommand):
+    """Implementation of 'clone' subcommand."""
+
+    CMD = 'clone'
+    EXAMPLE = "<dist-to-copy> <output-directory>"
+    HELP = "Write hod configs to a directory for editing purposes."
+
+    def run(self, args):
+        """Run 'genconfig' subcommand."""
```
Comment (boegel) [inline hod/subcommands/clone.py]: not genconfig

File: hod/subcommands/clone.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    """Option parser for 'clone' subcommand."""
+    VERSION = HOD_VERSION
+    ALLOPTSMANDATORY = False # let us use optionless arguments.
+
+
+class CloneSubCommand(SubCommand):
+    """Implementation of 'clone' subcommand."""
+
+    CMD = 'clone'
+    EXAMPLE = "<dist-to-copy> <output-directory>"
```
Comment (boegel) [inline hod/subcommands/clone.py]: `--dist`?

File: hod/subcommands/clone.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        optparser = CloneOptions(go_args=args, usage=self.usage_txt)
+        if len(optparser.args) < 3:
+            sys.stderr.write(self.usage())
+            sys.exit(1)
+
+        dist = optparser.args[1]
+        output_dir = optparser.args[2]
+
+        if os.path.exists(output_dir):
+            sys.stderr.write('Error: Output directory exists.\n')
```
Comment (boegel) [inline hod/subcommands/clone.py]: include output dir in error message

File: hod/subcommands/clone.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        try:
+            dists_dir = hcc.resolve_dists_dir()
+            src = os.path.join(dists_dir, dist)
+            if not os.path.exists(src):
+                sys.stderr.write('Error: Input dist not found: "%s".\n' % dist)
+                sys.exit(1)
+            shutil.copytree(src, output_dir)
+            return 0
+        except Exception as e:
+            _log.error("Failed to copy dist: %s", str(e))
```
Comment (boegel) [inline hod/subcommands/clone.py]: hmm, does this work with the formatting?

we usually do `_log.error("Failed to copy dist: %s" % err)`

**Comment by boegel** [comment]
@ehiggs: reviewed, small remarks

File: docs/Command_line_interface.rst (hunk 1 above)
Comment (ehiggs) [inline docs/Command_line_interface.rst]: The order is based on the amount of use we anticipate. So the most important and used functions are at the top. If you think it should change (e.g. genconfig and helptemplate should be lower) then we can file a new issue.

Comment (boegel) [inline docs/Command_line_interface.rst]: I think we've come to a point where we have too many subcommands to order them based on amount of use we anticipate. So, the only other alternative that makes sense is to just sort alphabetically (easier on the eyes too imho).

File: hod/subcommands/clone.py (hunk 6 above)
Comment (ehiggs) [inline hod/subcommands/clone.py]: We use positional arguments like `cp`, `mv`, `ln`, `scp`, `rsync`, etc.

File: hod/subcommands/clone.py (hunk 8 above)
Comment (ehiggs) [inline hod/subcommands/clone.py]: Yeah it works. I don't even need the `str` call apparently.

```python
In [2]: from vsc.utils import fancylogger
In [3]: _log = fancylogger.getLogger(fname=False)
In [4]: _log.error('hello darkness my old %s', 'friend')
2015-10-14 11:20:47,175 ERROR      ipython         MainThread  hello darkness my old friend
In [3]: _log.error('hello darkness my old %s', str(ValueError('friend')))
2015-10-14 11:22:28,992 ERROR      ipython         MainThread  hello darkness my old friend
In [4]: _log.error('hello darkness my old %s', ValueError('friend'))
2015-10-14 11:22:35,428 ERROR      ipython         MainThread  hello darkness my old friend
```

File: hod/subcommands/clone.py (hunk 4 above)
Comment (ehiggs) [inline hod/subcommands/clone.py]: Discussed off PR. We use like `cp`, `mv`, etc here. `hod connect` and `hod relabel` also use positional arguments.

Comment (boegel) [inline hod/subcommands/clone.py]: ok, less typing: good

**Comment by boegel** [comment]
Good to go, thanks @ehiggs!

**Pull request merged**